File: backend/src/main/java/com/autocare/vehicle/controller/VehicleController.java
package com.autocare.vehicle.controller;

import com.autocare.vehicle.dto.VehicleRequestDTO;
import com.autocare.vehicle.dto.VehicleResponseDTO;
import com.autocare.vehicle.service.VehicleService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/vehicles")
@RequiredArgsConstructor
public class VehicleController {

    private final VehicleService vehicleService;

    @PostMapping
    public ResponseEntity<VehicleResponseDTO> create(@Valid @RequestBody VehicleRequestDTO request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(vehicleService.create(request));
    }

    @GetMapping
    public ResponseEntity<List<VehicleResponseDTO>> findAll() {
        return ResponseEntity.ok(vehicleService.findAll());
    }

    @GetMapping("/{id}")
    public ResponseEntity<VehicleResponseDTO> findById(@PathVariable String id) {
        return ResponseEntity.ok(vehicleService.findByIdResponse(id));
    }

    @GetMapping("/client/{clientId}")
    public ResponseEntity<List<VehicleResponseDTO>> findByClient(@PathVariable String clientId) {
        return ResponseEntity.ok(vehicleService.findByClient(clientId));
    }

    @PutMapping("/{id}")
    public ResponseEntity<VehicleResponseDTO> update(
            @PathVariable String id,
            @Valid @RequestBody VehicleRequestDTO request) {
        return ResponseEntity.ok(vehicleService.update(id, request));
    }

    @PatchMapping("/{id}/mileage")
    public ResponseEntity<Void> updateMileage(
            @PathVariable String id,
            @RequestParam Integer mileage) {
        vehicleService.updateMileage(id, mileage);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> delete(@PathVariable String id) {
        vehicleService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/vehicle/vehicle.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  Patch,
  Post,
  Put,
  Query,
  UseGuards,
} from "@nestjs/common";
import { VehicleService } from "./vehicle.service";
import { VehicleRequestDto } from "./dto/vehicle.dto";
import { Roles } from "../common/decorators/roles.decorator";
import { RolesGuard } from "../common/guards/roles.guard";
import { Role } from "@prisma/client";

@Controller("vehicles")
@UseGuards(RolesGuard)
@Roles(Role.ADMIN, Role.RECEPTIONIST, Role.MANAGER, Role.MECHANIC)
export class VehicleController {
  constructor(private readonly vehicleService: VehicleService) {}

  @Post()
  @Roles(Role.ADMIN, Role.RECEPTIONIST)
  @HttpCode(HttpStatus.CREATED)
  create(@Body() dto: VehicleRequestDto) {
    return this.vehicleService.create(dto);
  }

  @Get()
  findAll() {
    return this.vehicleService.findAll();
  }

  @Get("client/:clientId")
  findByClient(@Param("clientId") clientId: string) {
    return this.vehicleService.findByClient(clientId);
  }

  @Get(":id")
  findById(@Param("id") id: string) {
    return this.vehicleService.findByIdResponse(id);
  }

  @Put(":id")
  @Roles(Role.ADMIN, Role.RECEPTIONIST)
  update(@Param("id") id: string, @Body() dto: VehicleRequestDto) {
    return this.vehicleService.update(id, dto);
  }

  @Patch(":id/mileage")
  @Roles(Role.ADMIN, Role.RECEPTIONIST, Role.MECHANIC)
  @HttpCode(HttpStatus.NO_CONTENT)
  async updateMileage(@Param("id") id: string, @Query("mileage", ParseIntPipe) mileage: number) {
    await this.vehicleService.updateMileage(id, mileage);
  }

  @Delete(":id")
  @Roles(Role.ADMIN, Role.RECEPTIONIST)
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param("id") id: string) {
    await this.vehicleService.delete(id);
  }
}
